File: kotlin/src/main/kotlin/380.InsertDeleteGetRandom.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/insert-delete-getrandom-o1/


class InsertDeleteGetRandom{
  class RandomizedSet() {
    /** Initialize your data structure here. */
    val indexMap = HashMap<Int, Int>()
    val values = Stack<Int>()
    val rand = Random()

    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    fun insert(v: Int): Boolean {
      if(indexMap.containsKey(v))
        return false
      values.add(v)
      indexMap[v] = values.size-1
      return true
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    fun remove(v: Int): Boolean {
      if(!indexMap.containsKey(v))
        return false
      val i = indexMap.remove(v)!!
      val last = values.pop()
      if(v != last){
        values[i] = last
        indexMap[last] = i
      }
      return true
    }

    /** Get a random element from the set. */
    fun getRandom(): Int {
      return values.get(rand.nextInt(values.size))
    }
  }

  private val randomSet = RandomizedSet()

  @Test
  fun test1(){
    assertEquals(true, randomSet.insert(1))
    assertEquals(false, randomSet.remove(2))
    assertEquals(true, randomSet.insert(2))
    assertEquals(true, randomSet.getRandom() in listOf(1,2))
    assertEquals(true, randomSet.remove(1))
    assertEquals(false, randomSet.insert(2))
    assertEquals(true, randomSet.getRandom() in listOf(2))
  }

}

File: python/380_InsertDeleteGetRandom.py
from unittest import TestCase
# https://leetcode.com/problems/insert-delete-getrandom-o1/
import random


class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.values = []
        self.index = {}

    def insert(self, val):
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        if self.index.get(val) is not None:
            return False
        self.index[val] = len(self.values)
        self.values.append(val)
        return True

    def remove(self, val):
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        :type val: int
        :rtype: bool
        """
        if self.index.get(val) is None:
            return False
        i = self.index.pop(val)
        last = self.values.pop()
        if val != last:
            self.index[last] = i
            self.values[i] = last
        return True

    def getRandom(self):
        """
        Get a random element from the set.
        :rtype: int
        """
        return random.choice(self.values)


class Tests(TestCase):
    def test1(self):
        randomSet = RandomizedSet()
        self.assertTrue(randomSet.insert(1))
        self.assertFalse(randomSet.remove(2))
        self.assertTrue(randomSet.insert(2))
        self.assertIn(randomSet.getRandom(), [1,2])
        self.assertTrue(randomSet.remove(1))
        self.assertFalse(randomSet.insert(2))
        self.assertIn(randomSet.getRandom(), [2])
        self.assertIn(randomSet.getRandom(), [2])
